src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Usage/DetectPLINQNopsAnalyzer.cs

```diff
@@ -0,0 +1,97 @@
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterCompilationStartAction(ctx =>
+            {
+                if (!ctx.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemLinqParallelEnumerable, out var parallelEnumerable))
+                {
+                    return;
+                }
+
+                var asParallelSymbols = parallelEnumerable.GetMembers("AsParallel").ToImmutableHashSet();
+                var toArraySymbols = parallelEnumerable.GetMembers("ToArray").ToImmutableHashSet();
+                var toListSymbols = parallelEnumerable.GetMembers("ToList").ToImmutableHashSet();
+
+                ctx.RegisterOperationAction(x => AnalyzeOperation(x, asParallelSymbols, toArraySymbols, toListSymbols), OperationKind.Invocation);
+            });
+        }
+
+        public static bool ParentIsForEachStatement(IInvocationOperation operation) => operation.Parent is IForEachLoopOperation || operation.Parent?.Parent is IForEachLoopOperation;
+
+        public static bool TryGetParentIsToArrayOrToList(IInvocationOperation operation, ImmutableHashSet<ISymbol> toArraySymbols, ImmutableHashSet<ISymbol> toListSymbols, out IInvocationOperation parentInvocation)
+        {
+            parentInvocation = operation;
+            if (operation.Parent?.Parent is not IInvocationOperation invocation)
+            {
+                return false;
+            }
+            if (toListSymbols.Contains(invocation.TargetMethod.OriginalDefinition) || toArraySymbols.Contains(invocation.TargetMethod.OriginalDefinition))
+            {
+                parentInvocation = invocation;
+                return true;
+            }
+            return false;
+        }
+
+        private static void AnalyzeOperation(OperationAnalysisContext context, ImmutableHashSet<ISymbol> asParallelSymbols, ImmutableHashSet<ISymbol> toArraySymbols, ImmutableHashSet<ISymbol> toListSymbols)
+        {
+            var invocation = (IInvocationOperation)context.Operation;
+            var reducedMethod = invocation.TargetMethod.OriginalDefinition;
+            if (reducedMethod is null)
+            {
+                return;
+            }
+
+            if (!asParallelSymbols.Contains(reducedMethod))
+            {
+                return;
+            }
+
+            IInvocationOperation? diagnosticInvocation = null;
+            if (!ParentIsForEachStatement(invocation))
+            {
+                if (!TryGetParentIsToArrayOrToList(invocation, toArraySymbols, toListSymbols, out var parentInvocation) || !ParentIsForEachStatement(parentInvocation))
+                {
+                    return;
+                }
+                diagnosticInvocation = parentInvocation;
+            }
+
+            diagnosticInvocation ??= invocation;
+            var diagnostic = diagnosticInvocation.CreateDiagnostic(DefaultRule, diagnosticInvocation.Syntax);
```

Comment:
Do we really want to call `diagnosticInvocation.Syntax.ToString()` for the extra arg?

Reply:
It could get long and ugly depending on the length of the foreach statement expression

Reply:
Agreed, the message should instead be `'AsParallel' in has no effect within a foreach statement`